packages/core/src/lib/filter/index.ts

```diff
@@ -113,64 +114,106 @@ class Subscription {
         FilterSubscribeResponse.decode(res[0].slice());
 
       if (statusCode < 200 || statusCode >= 300) {
-        throw new Error(
-          `Filter subscribe request ${requestId} failed with status code ${statusCode}: ${statusDesc}`
-        );
+        return {
+          requestId: requestId,
+          statusCode: statusCode,
+          message: statusDesc
+        };
       }
 
+      // Save the callback functions by content topics so they
+      // can easily be removed (reciprocally replaced) if `unsubscribe` (reciprocally `subscribe`)
+      // is called for those content topics
+      decodersGroupedByCT.forEach((decoders, contentTopic) => {
+        // Cast the type because a given `subscriptionCallbacks` map may hold
+        // Decoder that decode to different implementations of `IDecodedMessage`
+        const subscriptionCallback = {
+          decoders,
+          callback
+        } as unknown as SubscriptionCallback<IDecodedMessage>;
+
+        // The callback and decoder may override previous values, this is on
+        // purpose as the user may call `subscribe` to refresh the subscription
+        this.subscriptionCallbacks.set(contentTopic, subscriptionCallback);
+      });
+
       log.info(
         "Subscribed to peer ",
         this.peer.id.toString(),
         "for content topics",
         contentTopics
       );
+
+      return {
+        requestId: requestId,
+        statusCode: statusCode
+      };
     } catch (e) {
       throw new Error(
```

Comment:
There are still a few errors being thrown here and line 109. They should also be converted. Ideally nothing is throw and users entirely rely on the enum error